### PR TITLE
K01 ChargePoint refactorings

### PR DESCRIPTION
**Pull request by gberardi-pillar**
This feature branch includes work that updates the ChargePoint constructor to take references instead of shared pointers as parameters, as well as provides a MessageQueueInterface to allow for the creation of a MessageQueueMock, which will make it possible to run tests faster as well as write and maintain tests more easily. 

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [ ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: include/ocpp/common/message_queue.hpp [hunk 1]
```diff
@@ -661,16 +701,16 @@ template <typename M> class MessageQueue {
     }
 
     /// \brief Sends a new \p call_result message over the websocket
-    template <class T> void push(CallResult<T> call_result) {
+    void push_call_result(const json& call_result_json, const MessageId& unique_id) override {
```
Comment (shankari) [inline include/ocpp/common/message_queue.hpp:704]: so we don't need this to be a template any more because we just separate out the structure?

File: include/ocpp/common/message_queue.hpp [hunk 2]
```diff
@@ -966,77 +1004,59 @@ template <typename M> class MessageQueue {
...
-    bool contains_transaction_messages(const CiString<36> transaction_id) {
-        std::lock_guard<std::recursive_mutex> lk(this->message_mutex);
-        for (const auto control_message : this->transaction_message_queue) {
-            if (control_message->messageType == v201::MessageType::TransactionEvent) {
-                v201::TransactionEventRequest req = control_message->message.at(CALL_PAYLOAD);
-                if (req.transactionInfo.transactionId == transaction_id) {
-                    return true;
-                }
-            }
-        }
```
Comment (shankari) [inline include/ocpp/common/message_queue.hpp:991]: why is this removed?
The related change is
https://github.com/US-JOET/libocpp/commit/b72a6f601f62c8cdb80f6ffdd74f613ad12da4b4

which does not have any explanations

File: include/ocpp/common/message_queue.hpp [hunk 3]
```diff
@@ -966,77 +1004,59 @@ template <typename M> class MessageQueue {
...
-    bool contains_stop_transaction_message(const int32_t transaction_id) {
-        std::lock_guard<std::recursive_mutex> lk(this->message_mutex);
-        for (const auto control_message : this->transaction_message_queue) {
-            if (control_message->messageType == v16::MessageType::StopTransaction) {
-                v16::StopTransactionRequest req = control_message->message.at(CALL_PAYLOAD);
-                if (req.transactionId == transaction_id) {
-                    return true;
-                }
-            }
-        }
```
Comment (shankari) [inline include/ocpp/common/message_queue.hpp:1004]: Ditto

File: tests/lib/ocpp/v201/test_composite_schedule.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_composite_schedule.cpp:1]: The changes in this file don't appear to be related to the message queue.
Could you add a summary of the related changes?

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1]: Ditto. This doesn't seem to be related to the message queue at all, and I don't understand the motivation for the change.

**Review by shankari — CHANGES_REQUESTED** [review]
Again, a few high-level questions, for the record, that should be easy to answer.

File: include/ocpp/common/message_queue.hpp (hunk 2 above)
Comment (gberardi-pillar) [reply include/ocpp/common/message_queue.hpp:991]: It was not removed. It was moved into the v201 MessageQueue implementation (see lib/ocpp/v201/message_queue.cpp) since this logic is specific to v201::MessageType. Similarly with the v16::MessageType implementation which is in lib/ocpp/v16/message_queue.cpp.

File: tests/lib/ocpp/v201/test_composite_schedule.cpp
Comment (gberardi-pillar) [reply tests/lib/ocpp/v201/test_composite_schedule.cpp:1]: I'm not sure which changes you are referring to? 

The changes in this file are all related to the MessageQueue, specifically using a MessageQueueMock in tests which is what these changes make possible.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (gberardi-pillar) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1]: The other changes in this PR are related to making it possible to pass in parameters, such as the database_handler, to ChargePoint as references rather than as pointers.

File: include/ocpp/common/message_queue.hpp (hunk 1 above)
Comment (gberardi-pillar) [reply include/ocpp/common/message_queue.hpp:704]: The template part is now in the interface, which delegates to the derived class implementation, which does not need a template.

Basically, in C++ you cannot have a pure virtual function that is templated, yet we wanted a MessageQueueInterface to allow for the ability to mock it in tests. So some of the logic was moved into the interface template functions in order to create and pass the non-templated parameters to the derived implementation.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1]: In that case, it would have been helpful to flag that this PR contains two sets of unrelated changes 😄  and include the motivation for the second set of changes as well.

File: tests/lib/ocpp/v201/test_composite_schedule.cpp
Comment (shankari) [reply tests/lib/ocpp/v201/test_composite_schedule.cpp:1]: These changes
https://github.com/EVerest/libocpp/pull/831/files#diff-a117871b4a0e09ac30845bccb5123506c8c180dfa71f9c8abc8c61ae374dcdcbR207

which are clearly not related to the `MessageQueue`

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (gberardi-pillar) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1]: The PR description says "This feature branch includes work that updates the ChargePoint constructor to take references instead of shared pointers as parameters, as well as provides a MessageQueueInterface to allow for the creation of a MessageQueueMock,"

Is there something more I should have said to indicate that this includes two separate changes in a single PR?

**Review by shankari — APPROVED** [review]

**Comment by hikinggrass** [comment]
Closed in favor of #849 which will refactor the message dispatching. We will include parts of this PR either there or in another PR
